File: app/src/main/java/li/doerf/hacked/remote/haveibeenpwned/HIBPAccountResponseWorker.java
package li.doerf.hacked.remote.haveibeenpwned;

import android.app.Notification;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.app.NotificationCompat;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

import org.joda.time.DateTime;

import java.io.IOException;
import java.lang.ref.WeakReference;

import li.doerf.hacked.R;
import li.doerf.hacked.activities.MainActivity;
import li.doerf.hacked.db.AppDatabase;
import li.doerf.hacked.db.daos.AccountDao;
import li.doerf.hacked.db.daos.BreachDao;
import li.doerf.hacked.db.entities.Account;
import li.doerf.hacked.db.entities.Breach;
import li.doerf.hacked.utils.AccountHelper;
import li.doerf.hacked.utils.NotificationHelper;
import li.doerf.hacked.utils.OreoNotificationHelper;
import li.doerf.hacked.utils.StringHelper;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class HIBPAccountResponseWorker extends Worker {

    private static final String TAG = "HIBPAccountResponseWork";
    private static final String NOTIFICATION_GROUP_KEY_BREACHES = "group_key_breachs";
    public static final String KEY_ACCOUNT = "Account";
    public static final String KEY_BREACHES = "Breaches";
    private final WeakReference<Context> myContext;
    private final AccountDao myAccountDao;
    private final BreachDao myBreachDao;

    public HIBPAccountResponseWorker(@NonNull Context aContext, @NonNull WorkerParameters workerParams) {
        super(aContext, workerParams);
        myContext = new WeakReference<>(aContext);
        myAccountDao = AppDatabase.get(aContext).getAccountDao();
        myBreachDao = AppDatabase.get(aContext).getBreachDao();
    }

    @NonNull
    @Override
    public Result doWork() {
        String[] breaches = getInputData().getStringArray(KEY_BREACHES);
        String accountName = getInputData().getString(KEY_ACCOUNT);
        Account account = myAccountDao.findByName(accountName).get(0);

        try {
            boolean foundNewBreach  = false;

            for (String breachName : breaches) {
                Log.d(TAG, breachName);
                foundNewBreach |= handleBreach(account, breachName);
            }

            account.setLastChecked(DateTime.now());
            if (foundNewBreach && ! account.getHacked() ) {
                account.setHacked(true);
            }
            if (foundNewBreach) {
                new AccountHelper(myContext.get()).updateBreachCounts(account);
            }
            myAccountDao.update(account);


            // TODO v3 check if breach already exists
            // TODO v3 if not add breach

            return Result.success();
        } catch (IOException e) {
            Log.e(TAG, "caughtIOException while contacting www.haveibeenpwned.com - " + e.getMessage(), e);
            // TODO better error
            new Handler(Looper.getMainLooper()).post(() -> Toast.makeText(myContext.get(), myContext.get().getString(R.string.toast_error_error_during_check), Toast.LENGTH_LONG).show());
            return Result.failure();
        }
    }

    private boolean handleBreach(Account account, String breachName) throws IOException {
        Breach existing = myBreachDao.findByAccountAndName(account.getId(), breachName);

        if (existing != null) {
            Log.d(TAG, "breach already existing: " + breachName);
            return false;
        }

        BreachedAccount ba = getBreachedAccount(breachName);

        Log.d(TAG, "new breach: " + ba.getName());
        Breach newBreach = new Breach();
        newBreach.setAccount(account.getId());
        newBreach.setName(ba.getName());
        newBreach.setTitle(ba.getTitle());
        newBreach.setDomain(ba.getDomain());
        newBreach.setBreachDate(DateTime.parse(ba.getBreachDate()).getMillis());
        newBreach.setAddedDate(DateTime.parse(ba.getAddedDate()).getMillis());
        newBreach.setPwnCount(ba.getPwnCount());
        newBreach.setDescription(ba.getDescription());
        newBreach.setDataClasses(ba.getDataClasses() != null ? StringHelper.join(ba.getDataClasses(), ", ") : "");
        newBreach.setVerified(ba.getIsVerified());
        newBreach.setAcknowledged(false);
        myBreachDao.insert(newBreach);
        Log.i(TAG, "breach inserted into db");

        return true;
    }

    private BreachedAccount getBreachedAccount(String breachName) throws IOException {
        HttpLoggingInterceptor logging = new HttpLoggingInterceptor(message -> Log.v("OkHttp", message));
        // set your desired log level
        logging.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient.Builder httpClient = new OkHttpClient.Builder();
        httpClient.addInterceptor(logging);

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://haveibeenpwned.com/")
                .addConverterFactory(GsonConverterFactory.create())
                .client(httpClient.build())
                .build();

        HaveIBeenPwned service = retrofit.create(HaveIBeenPwned.class);
        Call<BreachedAccount> call = service.getBreach(breachName);
        Response<BreachedAccount> response = call.execute();
        return response.body();
    }

    private void doPostCheckActions(Boolean foundNewBreach) {
        // TODO here or before?
//        Intent localIntent = new Intent(HIBPAccountCheckerWorker.BROADCAST_ACTION_ACCOUNT_CHECK_FINISHED);
//        LocalBroadcastManager.getInstance(myContext.get()).sendBroadcast(localIntent);
//        Log.d(TAG, "broadcast finish sent");

        if ( foundNewBreach) {
            showNotification();
        }
    }

    private void showNotification() {
        if ( android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.O ) {
            OreoNotificationHelper onh = new OreoNotificationHelper(myContext.get());
            onh.createNotificationChannel();
        }

        String title = myContext.get().getString(R.string.notification_title_new_breaches_found);
        androidx.core.app.NotificationCompat.Builder mBuilder =
                new NotificationCompat.Builder(myContext.get(), OreoNotificationHelper.CHANNEL_ID)
                        .setSmallIcon(android.R.drawable.ic_dialog_info)
                        .setContentTitle(title)
                        .setContentText(myContext.get().getString(R.string.notification_text_click_to_open))
                        .setChannelId(OreoNotificationHelper.CHANNEL_ID)
                        .setOnlyAlertOnce(true)
                        .setGroup(NOTIFICATION_GROUP_KEY_BREACHES);

        Intent showBreachDetails = new Intent(myContext.get(), MainActivity.class);
        PendingIntent resultPendingIntent =
                PendingIntent.getActivity(
                        myContext.get(),
                        0,
                        showBreachDetails,
                        PendingIntent.FLAG_ONE_SHOT
                );
        mBuilder.setContentIntent(resultPendingIntent);

        Notification notification = mBuilder.build();
        NotificationHelper.notify(myContext.get(), notification);
    }
}

File: app/src/main/kotlin/li/doerf/hacked/remote/hibp/HIBPAccountResponseWorker.kt
package li.doerf.hacked.remote.hibp

import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.widget.Toast
import androidx.core.app.NotificationCompat
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.PropertyNamingStrategy
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import com.github.kittinunf.fuel.core.ResponseDeserializable
import com.github.kittinunf.fuel.core.isSuccessful
import com.github.kittinunf.fuel.coroutines.awaitObjectResponseResult
import com.github.kittinunf.fuel.httpGet
import com.google.firebase.crashlytics.FirebaseCrashlytics
import li.doerf.hacked.R
import li.doerf.hacked.activities.NavActivity
import li.doerf.hacked.db.AppDatabase
import li.doerf.hacked.db.daos.AccountDao
import li.doerf.hacked.db.daos.BreachDao
import li.doerf.hacked.db.entities.Account
import li.doerf.hacked.db.entities.Breach
import li.doerf.hacked.utils.AccountHelper
import li.doerf.hacked.utils.NotificationHelper
import li.doerf.hacked.utils.OreoNotificationHelper
import li.doerf.hacked.utils.StringHelper
import org.joda.time.DateTime
import java.io.IOException

class HIBPAccountResponseWorker(private val context: Context, workerParams: WorkerParameters) : CoroutineWorker(context, workerParams) {

    companion object {
        private const val TAG = "HIBPAccountResponseWork"
        private const val NOTIFICATION_GROUP_KEY_BREACHES = "group_key_breachs"
        const val KEY_ACCOUNT = "Account"
        const val KEY_BREACHES = "Breaches"
    }

    private val myAccountDao: AccountDao = AppDatabase.get(context).accountDao
    private val myBreachDao: BreachDao = AppDatabase.get(context).breachDao

    override suspend fun doWork(): Result {
        val breaches = inputData.getStringArray(KEY_BREACHES)
        val accountName = inputData.getString(KEY_ACCOUNT)
        val account = myAccountDao.findByName(accountName)[0]
        return try {
            var foundNewBreach = false
            for (breachName in breaches!!) {
                Log.d(TAG, breachName)
                foundNewBreach = foundNewBreach or handleBreach(account, breachName)
            }
            if (foundNewBreach) {
                showNotification()
            }
            updateAccount(account, foundNewBreach)
            Result.success()
        } catch (e: IOException) {
            Log.e(TAG, "caughtIOException while contacting www.haveibeenpwned.com - " + e.message, e)
            FirebaseCrashlytics.getInstance().recordException(e)
            // TODO better error
            Handler(Looper.getMainLooper()).post { Toast.makeText(context, context.getString(R.string.toast_error_error_during_check), Toast.LENGTH_LONG).show() }
            Result.failure()
        } catch (e: WorkFailedException) {
            if (e.retry) {
                Result.retry()
            } else {
                Result.failure()
            }
        } catch (e: Exception) {
            FirebaseCrashlytics.getInstance().recordException(e)
            throw e
        }
    }

    private fun updateAccount(account: Account, foundNewBreach: Boolean) {
        account.setLastChecked(DateTime.now())
        if (foundNewBreach && !account.hacked) {
            account.hacked = true
        }
        if (foundNewBreach) {
            AccountHelper(context).updateBreachCounts(account)
        }
        myAccountDao.update(account)
    }

    @Throws(IOException::class)
    private suspend fun handleBreach(account: Account, breachName: String): Boolean {
        val existing = myBreachDao.findByAccountAndName(account.id, breachName)
        if (existing != null) {
            Log.d(TAG, "breach already existing: $breachName")
            updateBreachedAccount(existing)
            return false
        }

        val ba = getBreachedAccount(breachName)
                ?: return false // request not successful

        addNewBreachedAccount(ba, account)
        return true
    }

    private suspend fun updateBreachedAccount(existing: Breach) {
        val lastCheckedDate = DateTime(existing.lastChecked)

        if (lastCheckedDate.isAfter(DateTime.now().minusWeeks(2))) {
            Log.d(TAG, "last check within last 14 days");
            return
        }

        val ba = getBreachedAccount(existing.name)
                ?: return // request not successful

        if (existing.modifiedDate == DateTime.parse(ba.modifiedDate).millis) {
            // no update to data
            existing.lastChecked = DateTime.now().millis
            myBreachDao.update(existing)
            Log.i(TAG, "existing breach last checked timestamp updated in db: ${existing.name}")
            return
        }

        existing.title = ba.title
        existing.domain = ba.domain
        existing.breachDate = DateTime.parse(ba.breachDate).millis
        existing.modifiedDate = DateTime.parse(ba.modifiedDate).millis
        existing.pwnCount = ba.pwnCount
        existing.description = ba.description
        existing.dataClasses = if (ba.dataClasses != null) StringHelper.join(ba.dataClasses, ", ") else ""
        existing.verified = ba.isVerified
        existing.fabricated = ba.isFabricated
        existing.retired = ba.isRetired
        existing.sensitive = ba.isSensitive
        existing.spamList = ba.IsSpamList
        existing.logoPath = ba.LogoPath
        existing.acknowledged = false
        existing.lastChecked = DateTime.now().millis
        myBreachDao.update(existing)
        Log.i(TAG, "existing breach updated in db: ${existing.name}")
    }

    private fun addNewBreachedAccount(ba: BreachedAccount, account: Account) {
        Log.d(TAG, "new breach: " + ba.name)
        val newBreach = Breach()
        newBreach.account = account.id
        newBreach.name = ba.name
        newBreach.title = ba.title
        newBreach.domain = ba.domain
        newBreach.breachDate = DateTime.parse(ba.breachDate).millis
        newBreach.addedDate = DateTime.parse(ba.addedDate).millis
        newBreach.modifiedDate = DateTime.parse(ba.modifiedDate).millis
        newBreach.pwnCount = ba.pwnCount
        newBreach.description = ba.description
        newBreach.dataClasses = if (ba.dataClasses != null) StringHelper.join(ba.dataClasses, ", ") else ""
        newBreach.verified = ba.isVerified
        newBreach.fabricated = ba.isFabricated
        newBreach.retired = ba.isRetired
        newBreach.sensitive = ba.isSensitive
        newBreach.spamList = ba.IsSpamList
        newBreach.logoPath = ba.LogoPath
        newBreach.acknowledged = false
        newBreach.lastChecked = DateTime.now().millis
        myBreachDao.insert(newBreach)
        Log.i(TAG, "breach inserted into db: " + newBreach.name)
    }

    @Throws(IOException::class)
    private suspend fun getBreachedAccount(breachName: String): BreachedAccount? {
        Log.d(TAG, "retrieving breaches for $breachName")
        val url = "https://haveibeenpwned.com/api/v3/breach/${breachName}"

        val (_, res, result) = url.httpGet()
                .header("Accept", "application/vnd.haveibeenpwned.v2+json")
                .header("User-Agent", "Hacked_Android_App")
                .awaitObjectResponseResult(BreachedAccountDeserializer)

        if (!res.isSuccessful) {
            Log.w(TAG, "request was not successful")
            return null
        }

        return result.get()
    }

    private fun showNotification() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val onh = OreoNotificationHelper(context)
            onh.createNotificationChannel()
        }
        val title = context.getString(R.string.notification_title_new_breaches_found)
        val mBuilder = NotificationCompat.Builder(context, OreoNotificationHelper.CHANNEL_ID)
                .setSmallIcon(android.R.drawable.ic_dialog_info)
                .setContentTitle(title)
                .setContentText(context.getString(R.string.notification_text_click_to_open))
                .setChannelId(OreoNotificationHelper.CHANNEL_ID)
                .setOnlyAlertOnce(true)
                .setGroup(NOTIFICATION_GROUP_KEY_BREACHES)
                .setAutoCancel(true)

        // TODO navigate directly to breaches
        val showBreachDetails = Intent(context, NavActivity::class.java)
        val resultPendingIntent = PendingIntent.getActivity(
                context,
                0,
                showBreachDetails,
                PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_ONE_SHOT
        )
        mBuilder.setContentIntent(resultPendingIntent)
        val notification = mBuilder.build()
        NotificationHelper.notify(context, notification)
    }

}

object BreachedAccountDeserializer : ResponseDeserializable<BreachedAccount> {
    override fun deserialize(content: String) = run {
        Log.d("BreachedAccountDeserial", content)
        val mapper = jacksonObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
        mapper.propertyNamingStrategy = PropertyNamingStrategy.UPPER_CAMEL_CASE
        mapper.readValue<BreachedAccount>(content)
    }
}